ADC_DMA/Core/Src/main.c

```diff
@@ -0,0 +1,189 @@
+	/*										*
+	 * --- ADC INITIALIZATION PROCEDURE --- *
+	 *										*/
+	RCC->APB2ENR 	|= 	RCC_APB2ENR_ADC1EN;		//enables ADC clock
+	ADC1->CR 		|=	ADC_CR_ADVREGEN;		//enables the voltage regulator
+	ADC1->CFGR2 	|= 	ADC_CFGR2_CKMODE_0;		//sets ADC clock as PCLK/2
+	ADC->CCR 		|= 	ADC_CCR_LFMEN;			//low frequency mode enable
+	ADC1->SMPR 		|= 	ADC_SMPR_SMPR_1	|		//sets sample time for 12.5 ADC clock cycles
+						ADC_SMPR_SMPR_0	;		//because we need 10.5 ADC clock cycles at least
+	ADC1->CFGR1		|=	ADC_CFGR1_DMACFG|		//DMA in circular mode
+						ADC_CFGR1_DMAEN	|		//enables DMA
+						ADC_CFGR1_WAIT;			//enables wait mode to prevent overrun
+
+	ADC1->CR |= ADC_CR_ADCAL;					//starting the calibration
+	while(ADC1->CR & ADC_CR_ADCAL);				//we have to wait until ADCAL = 0 (Can be handled by interrupt)
+
+	ADC1->ISR|= ADC_ISR_ADRDY;					//clear the ADRDY bit by programming it to 1
+	ADC1->CR |= ADC_CR_ADEN;					//then we enable the ADC
+	while(!(ADC1->ISR & ADC_ISR_ADRDY));		//and wait for it to be ready (Can be handled by interrupt)
+
+	/*												*
+	 * --- ADC CHANNEL CONFIG FOR V_REF READING --- *
+	 *												*/
+	if(ADC1->CR & ADC_CR_ADSTART)					//we have to be sure there's no ongoing conversion
+	{
+		ADC1->CR |= ADC_CR_ADSTP;					//if so, we have to set ADSTP bit
+		while(ADC1->CR & ADC_CR_ADSTP);				//and wait for it to clear
+	}
+	ADC1->CR |= ADC_CR_ADDIS;						//disables the ADC
+	while(ADC1->CR & ADC_CR_ADEN);
+
+	ADC1->CHSELR = ADC_CHSELR_CHSEL17;				//selecting the VREF channel
+	ADC->CCR |= ADC_CCR_VREFEN; 					//enables internal reference voltage
+
+	DMA1_Channel1->CNDTR = 1; 						//number of data to be transferred
+	DMA1_Channel1->CMAR = (uint32_t) &(v_ref);		//set memory register address for DMA
+	DMA1_Channel1->CCR 	|= DMA_CCR_EN;				//enables the DMA
+
+	ADC1->CR |= ADC_CR_ADEN;						//then we enable the ADC
+	while(!(ADC1->ISR & ADC_ISR_ADRDY));			//and wait for it to be ready. (Can be handled by interrupt)
+
+	/*									*
+	 * --- STARTING THE ADC READING --- *
+	 *									*/
+	ADC1->CR |= ADC_CR_ADSTART;								//starts the ADC
+
+	while(1)
+	{
+		while(!(ADC1->ISR & ADC_ISR_EOC)){}					//we wait for the ADC reading to finish
+		ADC1->ISR &= ~ADC_ISR_EOC;							//and clear the flag
+
+		//configuring the next channel to be read
+		DMA1_Channel1->CCR	&= ~DMA_CCR_EN;					//disables the DMA
+		ADC1->CHSELR = ADC_CHSELR_CHSEL18;					//selects internal temperature channel
+		ADC->CCR 	|= ADC_CCR_TSEN; 						//enables temperature sensor
+		wait(TIME_10uSEC);									//we have to wait for the proper time for the Tsense to wake up
+
+		//DMA1_Channel1->CNDTR = 1; 						//number of data to be transferred
+		DMA1_Channel1->CMAR = (uint32_t) &(temperature);	//set memory register address for DMA
+		DMA1_Channel1->CCR 	|= DMA_CCR_EN;					//enables the DMA
+
+		ADC1->CR 	|= ADC_CR_ADSTART;						//starts the ADC
+
+		while(!(ADC1->ISR & ADC_ISR_EOC)){}					//we wait for the ADC reading to finish
```

Comment:
Are you sure that that flag EOC also means that DMA is finished transferring the data?

Reply:
Well, the program works fine, and it looks to me that's how it was done in the TDR program.

There's DMA's transfer complete flag, that would be more reasonable to use, but if I change EOC to TCIF, or even if I use both, I'm not getting vref right

Reply:
ok, it's corrected now